lib/patch-eslint-scope.js

```diff
@@ -0,0 +1,376 @@
+    estraverse.VisitorKeys.ArrayPattern = ["elements"];
+    visitFunction.call(this, node);
+    // set them back to normal...
+    estraverse.VisitorKeys.ObjectPattern = t.VISITOR_KEYS.ObjectPattern;
+    estraverse.VisitorKeys.ArrayPattern = t.VISITOR_KEYS.ArrayPattern;
+    if (typeParamScope) {
+      this.close(node);
+    }
+  };
+
+  // visit flow type in VariableDeclaration
+  var variableDeclaration = referencer.prototype.VariableDeclaration;
+  referencer.prototype.VariableDeclaration = function(node) {
+    if (node.declarations) {
+      for (var i = 0; i < node.declarations.length; i++) {
+        var id = node.declarations[i].id;
+        var typeAnnotation = id.typeAnnotation;
+        if (typeAnnotation) {
+          checkIdentifierOrVisit.call(this, typeAnnotation);
+        }
+      }
+    }
+    variableDeclaration.call(this, node);
+  };
+
+  function createScopeVariable(node, name) {
+    this.currentScope().variableScope.__define(
+      name,
+      new Definition("Variable", name, node, null, null, null)
+    );
+  }
+
+  referencer.prototype.InterfaceDeclaration = function(node) {
+    createScopeVariable.call(this, node, node.id);
+    var typeParamScope;
+    if (node.typeParameters) {
+      typeParamScope = nestTypeParamScope.call(this, this.scopeManager, node);
+    }
+    // TODO: Handle mixins
+    for (var i = 0; i < node.extends.length; i++) {
+      visitTypeAnnotation.call(this, node.extends[i]);
+    }
+    visitTypeAnnotation.call(this, node.body);
+    if (typeParamScope) {
+      this.close(node);
+    }
+  };
+
+  referencer.prototype.TypeAlias = function(node) {
+    createScopeVariable.call(this, node, node.id);
+    var typeParamScope;
+    if (node.typeParameters) {
+      typeParamScope = nestTypeParamScope.call(this, this.scopeManager, node);
+    }
+    if (node.right) {
+      visitTypeAnnotation.call(this, node.right);
+    }
+    if (typeParamScope) {
+      this.close(node);
+    }
+  };
+
+  referencer.prototype.DeclareModule = referencer.prototype.DeclareFunction = referencer.prototype.DeclareVariable = referencer.prototype.DeclareClass = function(
+    node
+  ) {
+    if (node.id) {
+      createScopeVariable.call(this, node, node.id);
+    }
+
+    var typeParamScope;
+    if (node.typeParameters) {
+      typeParamScope = nestTypeParamScope.call(this, this.scopeManager, node);
+    }
+    if (typeParamScope) {
+      this.close(node);
+    }
+  };
+
+  referencer._babelEslintPatched = true;
+}
+
+// To patch for each call.
+var escope = null;
+var escopeAnalyze = null;
+
+module.exports = function(parserOptions) {
+  // Patch `Referencer.prototype` once.
+  if (!escope) {
+    try {
+      const modules = getModules();
+      monkeypatch(modules);
+
+      // Store to patch for each call.
+      escope = modules.escope;
+      escopeAnalyze = modules.escope.analyze;
+    } catch (err) {
+      console.error(err.stack);
+      process.exit(1);
```

Comment:
I don't think calling `process.exit(1)` here is a good idea, because another module could be doing something on the process. Throwing an error seems like it would be better instead.

Reply:
It's just copied from https://github.com/babel/babel-eslint/pull/559/commits/db14da01f2751053e1f911ad344c2966ea58e048#diff-3fa2324e5c70e869dd8b91ba47d140d1L379
But, I agree `process.exit(1)` is bad.

Reply:
Oh, I hadn't realized that. In that case I suppose we don't have to change it now.

Reply:
removed.